lib/Sema/CSSimplify.cpp

```diff
@@ -6354,18 +6360,38 @@ ConstraintSystem::SolutionKind ConstraintSystem::simplifyMemberConstraint(
   return SolutionKind::Error;
 }
 
-ConstraintSystem::SolutionKind
-ConstraintSystem::simplifyDefaultableConstraint(
-                                            Type first, Type second,
-                                            TypeMatchOptions flags,
-                                            ConstraintLocatorBuilder locator) {
+ConstraintSystem::SolutionKind ConstraintSystem::simplifyDefaultableConstraint(
+    Type first, Type second, TypeMatchOptions flags,
+    ConstraintLocatorBuilder locator) {
   first = getFixedTypeRecursive(first, flags, true);
 
   if (first->isTypeVariableOrMember()) {
     if (flags.contains(TMF_GenerateConstraints)) {
       addUnsolvedConstraint(
-        Constraint::create(*this, ConstraintKind::Defaultable, first, second,
-                           getConstraintLocator(locator)));
+          Constraint::create(*this, ConstraintKind::Defaultable, first, second,
+                             getConstraintLocator(locator)));
+      return SolutionKind::Solved;
+    }
+
+    return SolutionKind::Unsolved;
+  }
+
+  // Otherwise, any type is fine.
+  return SolutionKind::Solved;
+}
+
+ConstraintSystem::SolutionKind
+ConstraintSystem::simplifyDefaultClosureTypeConstraint(
+    Type closureType, Type inferredType,
+    ArrayRef<TypeVariableType *> referencedOuterParameters,
+    TypeMatchOptions flags, ConstraintLocatorBuilder locator) {
+  closureType = getFixedTypeRecursive(closureType, flags, /*wantRValue=*/true);
+
+  if (closureType->isTypeVariableOrMember()) {
```

Comment:
I wonder if it makes sense to add `hasTypeVariable()` here.

Reply:
I don't think so because we really want to keep this constraint until the closure is resolved and the type it's bound to can (and probably would) have type variables. This constraint serves dual purpose - source of type for the closure when there is not contextual information available e.g. `_ = { $0 + 1 }` and to connect closure type variable to its inferred parameters and result type + any outer parameters used in the body until the closure is resolved. When closure type variable becomes bound to inferred type (closure is "resolved") everything becomes connected, so `DefaultClosureType` is obsolete from that moment on. I was thinking instead of `DefaultCloureType` to go with `InferredClosureType`, maybe that's a better name...